### PR TITLE
AWS Init Support for Kuberflow/Testing

### LINKED ISSUE #748: Add some supports to make testing infra run on AWS
AWS plans to add some periodical tests for workshop https://github.com/aws-samples/eks-kubeflow-workshop. I think it makes more sense to setup this testing infra on AWS. I would like to leverage existing `kubeflow/testing` to test the repo and current blocking issue is some of the logics are GCP only. 

What we plan to do is still reusing the repo, add AWS utils, like `upload_to_s3`, `setup_iam_for_serviceaccount` etc, probably extract some common logic if needed and eventually make AWS an alternative platform to run testing infra. 

We'd like to put some time to help maintain common code base, help testing Tekton solution as well.  I am trying to get some feedbacks if this is a reasonable request and make sense to folks maintaining kubeflow/testing.

@Bobgy @jlewi @rmgogogo @pingsutw

### PR DESCRIPTION
**Pull request by PatrickXYS**
**Which issue is resolved by this Pull Request:**
Resolves #748  

**Description of your changes:**
1. Modify `run_workflows.sh` and `run_e2e_workflow.py` to make them AWS compatible.
2. Add `Dockerfile.py3.aws` for AWS ECR Image.
3. Add `images/aws-scripts` folder to place AWS related scripts.
4. Add `cloudprovider/aws` folder to place AWS related python files.
~~Since we added `import boto3` in `util.py`, GCP also need to install boto3 in the Dockerfile. Thus, modifies `Dockerfile` and `Dockerfile.py3`~~

### CODE REVIEW

File: images/Dockerfile.py3.aws
Comment (Jeffwan) [inline images/Dockerfile.py3.aws:94]: Let's remove this. auxiliary packages will be passed `checkout.sh`.  If we plan to reduce file downloads in the future, we can do it later in a separate story.

File: images/aws-scripts/check-deployment-status.sh
Comment (Jeffwan) [inline images/aws-scripts/check-deployment-status.sh:30]: I don't think we create ALB for all ephemeral cluster?

File: images/aws-scripts/check-deployment-status.sh
Comment (Jeffwan) [inline images/aws-scripts/check-deployment-status.sh:26]: If this file is needed, then fixed time should be changed to timeout and retry pattern.

File: images/aws-scripts/create-eks-cluster.sh
Comment (Jeffwan) [inline images/aws-scripts/create-eks-cluster.sh:23]: Most of the `presubmit` flow inject a `CLUSTER_NAME` env variable, have you checked period job?  Can we reuse that instead of bring a new env?

File: images/aws-scripts/create-eks-cluster.sh
Comment (Jeffwan) [inline images/aws-scripts/create-eks-cluster.sh:33]: Can you add some default value for this? 

```
MAX_NODE="${MIN_NODE:-3}"
```

File: images/aws-scripts/create-eks-cluster.sh
Comment (Jeffwan) [inline images/aws-scripts/create-eks-cluster.sh:26]: If you have TODO items, use `TODO`, don't use `AI` which is internally used term.

File: images/aws-scripts/delete-eks-cluster.sh
Comment (Jeffwan) [inline images/aws-scripts/delete-eks-cluster.sh:23]: same. if we can use `CLUSTER_NAME`

File: images/aws-scripts/deploy-kubeflow.sh
Comment (Jeffwan) [inline images/aws-scripts/deploy-kubeflow.sh:32]: Will this be changed later? Can you extract these into envs? If you have a follow up PR. I am fine to leave them here for now.

File: images/run_workflows.sh
Comment (Jeffwan) [inline images/run_workflows.sh:35]: why if `CLOUD_PROVIDER=gcp`?  Can you cover these edge cases?

File: py/kubeflow/testing/prow_artifacts.py
Comment (Jeffwan) [inline py/kubeflow/testing/prow_artifacts.py:228]: Can you use same layout?

Comment (Jeffwan) [reply py/kubeflow/testing/prow_artifacts.py:228]: Seems this method is not implemented

File: py/kubeflow/testing/prow_artifacts.py
Comment (Jeffwan) [inline py/kubeflow/testing/prow_artifacts.py:323]: Any reason leaving this empty?

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (Jeffwan) [inline py/kubeflow/testing/run_e2e_workflow.py:391]: Any reason not to use kf submit pipeline directly?

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (Jeffwan) [inline py/kubeflow/testing/run_e2e_workflow.py:660]: Can we get ride of these arguments? I think this makes `run_e2e_workflow.py` kind of heavy.

Alternatively, user can add envs to their steps to control the number of the nodes.

File: images/Dockerfile.py3.aws
Comment (PatrickXYS) [reply images/Dockerfile.py3.aws:14]: Yes, because those Google Packages are required in common python files.

Like `util.py` file, we not only have google packages but boto3, that means not only AWS but GCP's dockerfile needs to have all the necessary packages installed.

File: images/aws-scripts/check-deployment-status.sh
Comment (PatrickXYS) [reply images/aws-scripts/check-deployment-status.sh:30]: I think we create ALB, because we are using official kubeflow manifests, we enable ALB by default. Let me know if I'm wrong.

File: images/Dockerfile.py3.aws
Comment (PatrickXYS) [reply images/Dockerfile.py3.aws]: removed

File: images/Dockerfile.py3.aws
Comment (PatrickXYS) [reply images/Dockerfile.py3.aws]: Removed

**Review by andreyvelich — COMMENTED** [review]
@PatrickXYS Thanks a lot for doing this!
I left few comments

File: images/aws-scripts/check-deployment-status.sh
Comment (Jeffwan) [reply images/aws-scripts/check-deployment-status.sh:30]: I see. this would be used by jobs who create entire kubeflow stack like kfctl or manifest e2e. For most of the applications, they won't install entire kubeflow.  Can you update name to `check-load-balancer-status.sh`? 

`deployment` indicates checking Kubernetes Deployment status

File: images/Dockerfile.py3.aws
Comment (Jeffwan) [reply images/Dockerfile.py3.aws:65]: @andreyvelich I think the first step is to make sure all existing ksonnet jobs can work. It should be deprecated in a follow up story.

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (Jeffwan) [reply py/kubeflow/testing/run_e2e_workflow.py:486]: Yes. Argo and Tekton both should be supported.

File: images/Dockerfile.py3.aws
Comment (andreyvelich) [reply images/Dockerfile.py3.aws:65]: Got, it. Thanks @Jeffwan.

File: images/aws-scripts/check-deployment-status.sh
Comment (PatrickXYS) [reply images/aws-scripts/check-deployment-status.sh:26]: Good point, fix in next commit

File: images/aws-scripts/create-eks-cluster.sh
Comment (PatrickXYS) [reply images/aws-scripts/create-eks-cluster.sh:23]: Good point, fix in next commit

File: images/aws-scripts/create-eks-cluster.sh
Comment (PatrickXYS) [reply images/aws-scripts/create-eks-cluster.sh:33]: Good point, fix in next commit

File: images/aws-scripts/delete-eks-cluster.sh
Comment (PatrickXYS) [reply images/aws-scripts/delete-eks-cluster.sh:23]: Good point, fix in next commit

File: images/Dockerfile.py3.aws
Comment (PatrickXYS) [reply images/Dockerfile.py3.aws:65]: We'll start with ksonnet for now, that can makes existing test cases easy to be used in AWS test infra. 

But yes we'll meanwhile add support for Tekton, and gradually move to Tekton.

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py:486]: Yes and I think for now we use ksonnet as a start-point. It can help us leverage upstream existing tests.

Meanwhile we'll add support for Tekton in AWS.

File: images/run_workflows.sh
Comment (PatrickXYS) [reply images/run_workflows.sh:35]: The point I didn't add `CLOUD_PROVIDER=gcp` is it require GCP folks to add `ENV CLOUD_PROVIDER=gcp` in their Dockerfiles.

Since it's AWS related PR, it might does not make sense to ask GCP to make changes. 

I can add them if @jlewi think it's okay to add so.

File: images/aws-scripts/check-deployment-status.sh
Comment (PatrickXYS) [reply images/aws-scripts/check-deployment-status.sh:30]: You're right, will fix in new commit

File: images/aws-scripts/deploy-kubeflow.sh
Comment (PatrickXYS) [reply images/aws-scripts/deploy-kubeflow.sh:32]: We might need to come up with a better plan to add Compatibility matrix later.

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py:660]: Yeah it makes sense, will fix in new commit

File: images/aws-scripts/create-eks-cluster.sh
Comment (PatrickXYS) [reply images/aws-scripts/create-eks-cluster.sh:29]: Yes, will fix in new commit

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py]: I removed these params and user can configure them directly from ENVS.

File: images/aws-scripts/create-eks-cluster.sh
Comment (PatrickXYS) [reply images/aws-scripts/create-eks-cluster.sh]: Fixed

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py]: Fixed

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py]: I was thinking about if this is necessary to add another lay, because it's easy for developers to see the mappings between cloud_provider and "aws" / "gcp" / "ibm"

File: images/run_workflows.sh
Comment (jlewi) [reply images/run_workflows.sh:32]: Can we trigger the GCP logic if CLOUD_PROVIDER is not set or if it is gcp?

File: py/kubeflow/testing/prow_artifacts.py
Comment (PatrickXYS) [reply py/kubeflow/testing/prow_artifacts.py]: Fixed

File: py/kubeflow/testing/prow_artifacts.py
Comment (PatrickXYS) [reply py/kubeflow/testing/prow_artifacts.py:238]: Fixed in latest commit

File: images/run_workflows.sh
Comment (PatrickXYS) [reply images/run_workflows.sh:32]: Add below logic to handle GCP request:

```
if [ -z "$CLOUD_PROVIDER" ] ||  [ "$CLOUD_PROVIDER" = "gcp" ]
then
  python -m kubeflow.testing.run_e2e_workflow \
    --project=kubeflow-ci \
    --zone=us-east1-d \
    --cluster=kubeflow-testing \
    --bucket=kubernetes-jenkins \
    --config_file=/src/${REPO_OWNER}/${REPO_NAME}/prow_config.yaml \
    --repos_dir=/src
else

```

```
if not args.cloud_provider or args.cloud_provider == "gcp":
....
elif args.cloud_provider == "aws":
.... 
```

**Comment by PatrickXYS** [comment]
To make kubeflow-testing-presubmit succeed, we need GCP folks to first send PR to build new container image with `boto3` installed.

This is TBD for now. Will pick this up once we all agree to move forward and merge this PR.

**Comment by jlewi** [comment]
> To make kubeflow-testing-presubmit succeed, we need GCP folks to first send PR to build new container image with boto3 installed.

The folks listed in: https://github.com/kubeflow/internal-acls/blob/master/google_groups/ci-team.members.txt

Should be able to help with that

**Comment by PatrickXYS** [comment]
@jlewi Thanks for suggestions, made a new folder `cloudprovider/aws` to use any python scripts/functions.

Thus, the presubmit job succeeded.

Can you take another look? @Jeffwan @andreyvelich

**Comment by jlewi** [comment]
I recommend creating OWNERs files for all AWS related assets.
You can either organize these into subfolders or use [filters](https://github.com/kubernetes/community/blob/master/contributors/guide/owners.md#filters) to add different OWNERs for different/specific files in a directory (e.g. specify owners for your Dockerfile)

**Comment by PatrickXYS** [comment]
Given some of the files interact with GCP and AWS, like `run_workflows.sh`, `run_e2e_workflows.py`, I would like to create OWNERs file for both of them, and listed GCP and AWS folks as approvers.

**Comment by jlewi** [comment]
@Bobgy what do you think about @PatrickXYS comment about adding AWS folks to the OWNERs files. Given #737 would it better to cleanly separate the AWS and non-AWS test infra and minimize shared code?

@PatrickXYS Is anything other than the OWNERs file changes holding up this PR?

File: py/kubeflow/testing/OWNERS
Comment (jlewi) [inline py/kubeflow/testing/OWNERS:3]: Per my comment I think it might be better to move in the direction of better separation. With that in mind it might be better not to add additional OWNERs here. At the very least might be better to do this in a follow on PR to avoid blocking this PR.

Comment (PatrickXYS) [reply py/kubeflow/testing/OWNERS:3]: Make sense, will do this in a follow-up PR

**Comment by PatrickXYS** [comment]
@Jeffwan @jlewi @andreyvelich 

I cleaned up the code and nothing blocks the PR for now.

Can you take another look?

**Comment by andreyvelich** [comment]
@PatrickXYS Since we create another util for AWS, my suggestion is: left only functions that we are using.

So you can delete everything except these functions: `upload_to_s3`, `aws_configure_credential`, `load_kube_config`, `upload_file_to_s3`, `run`, `to_s3_uri`, `split_s3_uri`. 
Later we can add more features, if it is needed.

What do you think?

**Comment by jlewi** [comment]
On my end 
/lgtm

@Bobgy Thoughts?

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py:392]: Yeah let me clean it up

File: py/kubeflow/testing/run_e2e_workflow.py
Comment (PatrickXYS) [reply py/kubeflow/testing/run_e2e_workflow.py:370]: We don't have to, add it for AWS condition

File: images/aws-scripts/deploy-kubeflow.sh
Comment (PatrickXYS) [reply images/aws-scripts/deploy-kubeflow.sh:49]: We don't have to, the reason I did it here is to follow AWS Kubeflow Installation steps.

File: images/aws-scripts/uninstall-kubeflow.sh
Comment (PatrickXYS) [reply images/aws-scripts/uninstall-kubeflow.sh:36]: Not really

**Review by Jeffwan — APPROVED** [review]

**Review by andreyvelich — APPROVED** [review]
Thanks @PatrickXYS 
/lgtm

**Comment by Bobgy** [comment]
/approve
I don't have any particular objections, I'm not very familiar with kf testing infra

**Comment by PatrickXYS** [comment]
False-positive signal for presubmit, it actually succeed.

http://testing-argo.kubeflow.org/workflows/kubeflow-test-infra/kubeflow-testing-presubmit-py-unittests-755-0c30576-5072-7695?tab=workflow

**Pull request merged**
